sdk/keyvault/Azure.Security.KeyVault.Certificates/src/PemReader.cs

```diff
@@ -0,0 +1,191 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Reflection;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using Azure.Core;
+
+namespace Azure.Security.KeyVault.Certificates
+{
+    internal static class PemReader
+    {
+        private delegate void ImportPkcs8PrivateKeyDelegate(ReadOnlySpan<byte> blob, out int bytesRead);
+
+        private const string Prolog = "-----BEGIN ";
+        private const string Epilog = "-----END ";
+        private const string LabelEnd = "-----";
+
+        private static MethodInfo s_importPkcs8PrivateKeyMethod;
+        private static MethodInfo s_copyWithPrivateKeyMethod;
+
+        /// <summary>
+        /// Loads an <see cref="X509Certificate2"/> from PEM data.
+        /// </summary>
+        /// <param name="data">The PEM data to parse.</param>
+        /// <param name="cer">Optional public certificate data if not defined within the PEM data.</param>
+        /// <param name="allowCertificateOnly">Whether to create an <see cref="X509Certificate2"/> if no private key is read.</param>
+        /// <returns>An <see cref="X509Certificate2"/> loaded from the PEM data.</returns>
+        /// <exception cref="CryptographicException">A cryptographic exception occurred when trying to create the <see cref="X509Certificate2"/>.</exception>
+        /// <exception cref="InvalidOperationException"><paramref name="cer"/> is null and no CERTIFICATE field is defined in PEM, or no PRIVATE KEY is defined in PEM.</exception>
+        /// <exception cref="PlatformNotSupportedException">Creating a <see cref="X509Certificate2"/> from PEM data is not supported on the current platform.</exception>
+        public static X509Certificate2 LoadCertificate(ReadOnlySpan<char> data, byte[] cer = null, bool allowCertificateOnly = false)
+        {
+            byte[] priv = null;
+
+            while (TryRead(data, out PemField field))
+            {
+                if (field.Label.Equals("CERTIFICATE".AsSpan(), StringComparison.Ordinal) && cer is null)
+                {
+                    cer = field.FromBase64Data();
+                    data = data.Slice(field.Start + field.Length);
+                }
+                else if (field.Label.Equals("PRIVATE KEY".AsSpan(), StringComparison.Ordinal))
+                {
+                    priv = field.FromBase64Data();
+                }
+
+                int offset = field.Start + field.Length;
+                if (offset >= data.Length)
+                {
+                    break;
+                }
+
+                data = data.Slice(offset);
+            }
+
+            if (cer is null)
+            {
+                throw new InvalidOperationException("Missing public key");
```

Comment:
Should we add a bit more text to the message?

`The provided certificate data is missing public key`
?